src/workerd/api/streams/readable-source.c++

```diff
@@ -0,0 +1,817 @@
+    //
+    // Not everything can be best case scenario tho, unfortunately. If our first tryRead
+    // does not fully consume the stream or fully fill the destination buffer, we're
+    // going to need to try again. It is possible that the new allocation in the next
+    // iteration will be wasted if the stream doesn't have any more data so it's important
+    // for us to try to be conservative with the allocation. If the running total of data
+    // we've seen so far is equal to or greater than the expected total length of the stream,
+    // then the most likely case is that the next read will be zero-length -- but unfortunately
+    // we can't know for sure! So for this we will fall back to a more conservative allocation
+    // which is either 4096 bytes or the calculated amountToRead, whichever is the lower number.
+
+    kj::Vector<kj::Array<T>> parts;
+    size_t runningTotal = 0;
+    static constexpr size_t MIN_BUFFER_CHUNK = 1024;
+    static constexpr size_t DEFAULT_BUFFER_CHUNK = 4096;
+    static constexpr size_t MAX_BUFFER_CHUNK = DEFAULT_BUFFER_CHUNK * 4;
+
+    // If we know in advance how much data we'll be reading, then we can attempt to optimize the
+    // loop here by setting the value specifically so we are only allocating at most twice. But,
+    // to be safe, let's enforce an upper bound on each allocation even if we do know the total.
+    kj::Maybe<size_t> maybeLength = input.tryGetLength(rpc::StreamEncoding::IDENTITY);
+
+    // The amountToRead is the regular allocation size we'll use right up until we've read the
+    // number of expected bytes (if known). This number is calculated as the minimum of (limit,
+    // MAX_BUFFER_CHUNK, maybeLength or DEFAULT_BUFFER_CHUNK). In the best case scenario, this
+    // number is calculated such that we can read the entire stream in one go if the amount of
+    // data is small enough and the stream is well behaved.
+    //
+    // If the stream does report a length, once we've read that number of bytes, we'll
+    // fallback to the more conservative allocation.
+    size_t amountToRead =
+        kj::min(limit, kj::min(MAX_BUFFER_CHUNK, maybeLength.orDefault(DEFAULT_BUFFER_CHUNK)));
+
+    // amountToRead can be zero if the stream reported a zero-length. While the stream could
+    // be lying about its length, let's skip reading anything in this case.
+    if (amountToRead != 0) {
+      while (true) {
+        auto bytes = kj::heapArray<T>(amountToRead);
+        // Note that we're passing amountToRead as the *minBytes* here so the tryRead should
+        // attempt to fill the entire buffer. If it doesn't, the implication is that we read
+        // everything.
+        size_t amount = co_await input.read(bytes.asBytes(), amountToRead);
+        KJ_DASSERT(amount <= amountToRead);
+
+        runningTotal += amount;
+        JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
```

Comment:
Should be `<=` instead?

`amountToRead == limit` is possible (line 109-110), which means `amount == limit` is possible, which means `runningTotal == limit` is possible.

If so, there might be a missing test case. Try a `limit` of 1 with a stream that advertises a length of 1 (or none at all), and produces 1 byte before EOF.

Reply:
Yep, should be <= limit.